File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack();
        int maxArea = 0;
        int n = heights.length;
        
        for (int i = 0; i <= n; i++) {
            int height = (i == n) ? 0 : heights[i];
            while (!stack.isEmpty() && heights[stack.peek()] > height) {
                int h = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, h * width);
            }
            stack.push(i);
        }
        return maxArea;
    }
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # put heights in the stack in non-decreasing order
        stack, largest = [], 0
        for i, height in enumerate(heights + [0]):
            prev_i = i      # prev_i is imp, extend cur to all the way back
            while stack and stack[-1][0] > height:
                prev_h, prev_i = stack.pop()
                largest = max(largest, prev_h * (i - prev_i))
            stack.append((height, prev_i)) 
        return largest